src/nanoarrow/array.c

```diff
@@ -995,6 +1021,31 @@ static int ArrowArrayViewValidateDefault(struct ArrowArrayView* array_view,
         }
       }
       break;
+
+    case NANOARROW_TYPE_RUN_END_ENCODED: {
+      struct ArrowArrayView* run_ends_view;
+      run_ends_view = array_view->children[0];
+      if (run_ends_view->null_count != 0) {
+        ArrowErrorSet(error, "Run Ends cannot be null but found null_count value %ld",
+                      (long)run_ends_view->null_count);
+        return EINVAL;
+      }
+      for (int64_t i = 0, prev_run_end = 0; i < run_ends_view->length; i++) {
+        int64_t run_end = ArrowArrayViewGetIntUnsafe(run_ends_view, i);
+        if (run_end < 0 || run_end < prev_run_end) {
+          return EINVAL;
+        }
+        prev_run_end = run_end;
+        if (i == run_ends_view->length - 1 && run_end != array_view->length) {
+          ArrowErrorSet(error,
+                        "Run End value %ld at index %ld exceeds the logical "
+                        "length of the run-end encoded array %ld",
+                        (long)run_end, (long)i, (long)array_view->length);
+          return EINVAL;
+        }
+      }
```

Comment:
This is how I wrote this loop with a single branch inside the loop:

```cpp
      if (run_ends[0] < 1) {
        return Status::Invalid(
            "All run ends must be greater than 0 but the first run end is ", run_ends[0]);
      }
      int64_t last_run_end = run_ends[0];
      for (int64_t index = 1; index < run_ends_length; index++) {
        const int64_t run_end = run_ends[index];
        if (run_end <= last_run_end) {
          return Status::Invalid(
              "Every run end must be strictly greater than the previous run end, "
              "but run_ends[",
              index, "] is ", run_end, " and run_ends[", index - 1, "] is ",
              last_run_end);
        }
        last_run_end = run_end;
      }
```

https://github.com/apache/arrow/blob/e4baf6be2167eb6ccbda90275304336f49998eac/cpp/src/arrow/array/validate.cc#L753